File: cpp/src/controllers/AppController.cpp
#include "AppController.h"
#include "NotebookController.h"
#include <QNetworkReply>

AppController::AppController(QObject* parent)
    : QObject(parent)
    , m_restClient(new RestClient(this))
    , m_tracker(new ExecutionTracker(this))
{
}

AppController::~AppController() {
    qDeleteAll(m_notebooks);
}

void AppController::setServerConfig(const ServerConfig& config) {
    m_config = config;
    m_restClient->setConfig(config);

    // Seed XSRF cookie
    auto* reply = m_restClient->fetchXsrf();
    connect(reply, &QNetworkReply::finished, reply, &QNetworkReply::deleteLater);
}

void AppController::checkServer() {
    auto* reply = m_restClient->checkServer();
    connect(reply, &QNetworkReply::finished, this, [this, reply]() {
        reply->deleteLater();
        int code = reply->attribute(QNetworkRequest::HttpStatusCodeAttribute).toInt();
        if (reply->error() == QNetworkReply::NoError && code == 200) {
            emit serverChecked(QStringLiteral("ok"));
        } else if (code == 401 || code == 403) {
            emit serverChecked(QStringLiteral("unauthorized"));
        } else {
            emit serverChecked(reply->errorString());
        }
    });
}

void AppController::openNotebook(const QString& path) {
    // Check if already open
    for (auto it = m_notebooks.begin(); it != m_notebooks.end(); ++it) {
        if (it.value()->notebookModel()->path() == path) {
            emit notebookOpened(it.key(), it.value());
            return;
        }
    }

    QString notebookId = QStringLiteral("nb-%1").arg(++m_notebookCounter);
    auto* controller = new NotebookController(notebookId, m_restClient, m_tracker, this);
    m_notebooks[notebookId] = controller;

    controller->load(path, m_config);
    emit notebookOpened(notebookId, controller);
}

void AppController::closeNotebook(const QString& notebookId) {
    auto* controller = m_notebooks.take(notebookId);
    if (!controller) return;

    controller->shutdownKernel();
    controller->deleteLater();
    emit notebookClosed(notebookId);
}

NotebookController* AppController::notebookController(const QString& notebookId) const {
    return m_notebooks.value(notebookId, nullptr);
}

QList<NotebookController*> AppController::allNotebooks() const {
    return m_notebooks.values();
}

File: jupyterqt/controllers/app_controller.py
from PySide6.QtCore import QObject, Signal, QThreadPool

from jupyterqt.config import ServerConfig
from jupyterqt.controllers.notebook_controller import NotebookController
from jupyterqt.network.rest_client import RestClient
from jupyterqt.network.rest_workers import RestWorker


class AppController(QObject):
    notebook_opened = Signal(str, object)    # notebookId, NotebookController
    notebook_closed = Signal(str)            # notebookId
    notebook_created = Signal()              # fired after a new notebook file is created
    server_checked = Signal(str)             # "ok", "unauthorized", or error string

    def __init__(self, config: ServerConfig, parent=None):
        super().__init__(parent)
        self._config = config
        self._rest = RestClient(config)
        self._notebooks: dict[str, NotebookController] = {}
        self._active_notebook: NotebookController | None = None
        self._pool = QThreadPool.globalInstance()


    # #########################################################################################################################################
    #

    @property
    def config(self) -> ServerConfig:
        return self._config

    def updateConfig(self, config: ServerConfig) -> None:
        self._config = config
        self._rest.updateConfig(config)

    def checkServer(self) -> None:
        w = RestWorker(self._rest.checkServer)
        w.signals.finished.connect(lambda result: self.server_checked.emit(str(result)))
        w.signals.error.connect(lambda e: self.server_checked.emit(f"error: {e}"))
        self._pool.start(w)

    def openNotebook(self, path: str) -> None:
        # Avoid opening same path twice
        for ctrl in self._notebooks.values():
            if ctrl.path == path:
                self.notebook_opened.emit(ctrl.notebookId, ctrl)
                return

        ctrl = NotebookController(path, self._config, self)
        # We need notebookId after load; connect a one-shot
        ctrl.notebook_loaded.connect(lambda: self._onControllerLoaded(ctrl))
        ctrl.notebook_load_failed.connect(lambda e: self._onLoadFailed(ctrl, e))
        ctrl.load()

    def _onControllerLoaded(self, ctrl: NotebookController) -> None:
        notebookId = ctrl.notebookId
        self._notebooks[notebookId] = ctrl
        self.notebook_opened.emit(notebookId, ctrl)

    def _onLoadFailed(self, ctrl: NotebookController, error: str) -> None:
        ctrl.deleteLater()

    def closeNotebook(self, notebookId: str) -> None:
        ctrl = self._notebooks.pop(notebookId, None)
        if ctrl:
            if self._active_notebook is ctrl:
                self._active_notebook = None
            ctrl.cleanup()
            ctrl.deleteLater()
        self.notebook_closed.emit(notebookId)

    def getController(self, notebookId: str) -> NotebookController | None:
        return self._notebooks.get(notebookId)

    def createNotebook(self, directory: str = "") -> None:
        w = RestWorker(self._rest.createNotebook, directory)
        w.signals.finished.connect(self._onNotebookCreated)
        self._pool.start(w)

    def _onNotebookCreated(self, data: object) -> None:
        if isinstance(data, dict) and "path" in data:
            self.notebook_created.emit()
            self.openNotebook(data["path"])

    def activeNotebook(self) -> NotebookController | None:
        return self._active_notebook

    def setActiveNotebook(self, ctrl: NotebookController | None) -> None:
        print(f'Notebook {ctrl.path} is now active.')
        self._active_notebook = ctrl

    def allNotebooks(self) -> list[NotebookController]:
        return list(self._notebooks.values())
